x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java

```diff
@@ -218,6 +223,64 @@ public void startDeployment(TrainedModelDeploymentTask task, ActionListener<Trai
         );
     }
 
+    private void validateMLNodeArchitecturesAndModelMetadata(
+        Map<String, Object> metadata,
+        ActionListener<TrainedModelDeploymentTask> failedDeploymentListener
+    ) {
+        if (metadata != null) {
+            String modelPlatformArchitecture = ((String) metadata.get(TrainedModelConfig.PLATFORM_ARCHITECTURE));
+            if (modelPlatformArchitecture != null) {
+                verifyMLNodeArchitectureMatchesModelPlatformArchitecture(failedDeploymentListener, modelPlatformArchitecture);
+            }
+        }
+    }
+
+    private void verifyMLNodeArchitectureMatchesModelPlatformArchitecture(
+        ActionListener<TrainedModelDeploymentTask> failedDeploymentListener,
+        String modelPlatformArchitecture
+    ) {
+        ActionListener<Set<String>> architectureValidationListener = ActionListener.wrap(architectures -> {
+            try {
+                verifyArchitectureMatchesModelPlatformArchitecture(architectures, modelPlatformArchitecture);
+            } catch (IllegalArgumentException iae) {
+                verifyArchitectureOfMLNodesIsHomogenous(architectures);
+                throw iae;
+            }
+        }, failedDeploymentListener::onFailure);
+
+        SupportedPlatformArchitectures.getNodesOsArchitectures(threadPool, client, architectureValidationListener);
+    }
+
+    private void verifyArchitectureOfMLNodesIsHomogenous(Set<String> architectures) throws IllegalStateException {
+        if (architectures.size() > 1) {
+            throw new IllegalStateException(
+                format(
+                    "ML nodes in this cluster have multiple platform architectures, but can only have one; " + "was [%s];" + " %s",
+                    architectures.toString(),
+                    "<link to documentation?>"
+                )
+            );
+        }
+    }
+
+    private void verifyArchitectureMatchesModelPlatformArchitecture(Set<String> architectures, String modelPlatformArchitecture)
+        throws IllegalArgumentException {
+
+        String architecture = architectures.iterator().next();
```

Comment:
How should we handle this situation? I can see that we don't want to throw a NoSuchElementException, but if there aren't any ML nodes in the cluster, then it is true that the model architecture doesn't match the cluster. Perhaps its still just best to just throw the below IllegalArgumentException? Or perhaps we expect this type of situation to be caught somewhere else before this could would be executed?

Reply:
I think in the case of zero ML nodes we'll just have to allow any architecture to be installed and started.

The main case where zero ML nodes will occur is in Cloud with autoscaling. In that case the nodes will be `linux-x86_64` and we know our first platform-specific models will be for that architecture, so things will work OK.

On-prem users will almost never have autoscaling. The exception might be ECE users, but they're definitely on Linux and highly likely `x86_64`, so we should be OK there too.

So just add an extra test at the beginning of the method that says if there are zero ML nodes then the validation passes.